Test to configure rbd compression hint settings at multiple levels

sunilangadi2 opened the pull request:
# Description
Automated below test case: Test to verify "rbd_compression_hint" configs in globally, per-pool, or per-image
[CEPH-83574644](https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83574644)

Test flow:
1. Create a pool and an Image, write some data on it
2. set bluestore_compression_mode to passive to enable rbd_compression_hint feature
3. verify "rbd_compression_hint" to "compressible" on global, pool and image level
4. verify "rbd_compression_hint" to "incompressible" on global, pool and image level
5. Repeat the above steps for ecpool

Test result:
[Passed] http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-46AKM3

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

rkhambad commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,133 @@
+        int: The return value. 0 for success, 1 for failure.
+    """
+    log.info(
+        "Starting CEPH-83574644 - Validate rbd_compression_hint config"
```
give a space after config.

sunilangadi2 replied:
Done

rkhambad commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,133 @@
+
+        log.info("Able to set bluestore_compression_mode to passive")
+
+        for option in ["compressible", "incompressible"]:
```
[optional] we are verifying set and get twice on global level, we can think to run on global only once.

sunilangadi2 replied:
I don't think so, looks like you got confused with `entity` global level has entity of type global.

rkhambad commented on tests/rbd/rbd_utils.py:
```diff
@@ -773,6 +773,35 @@ def set_config(self, level, entity, key, value, **kw):
         """
         return self.exec_cmd(cmd=f"rbd config {level} set {entity} {key} {value}", **kw)
 
+    def set_ceph_config(self, entity, key, value, **kw):
```
[optional] we can think of modifying the existing set_config and get_config method, according to the parameter passed in function.

sunilangadi2 replied:
i feel it's good to keep `ceph` commands separate over `rbd` commands hence added this method.

manasagowri replied:
I agree with Sunil here, let's not combine rbd config and ceph config.

manasagowri replied:
But @sunilangadi2 set_ceph_config and get_ceph_config are generic methods not related to RBD object, so they can be moved out of the RBD class.

sunilangadi2 replied:
Done

sunilangadi2 commented:
Added validation steps at each config level,
Test result:[Passed]
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-6MUQGX/

manasagowri commented on tests/rbd/rbd_utils.py:
```diff
@@ -773,6 +773,35 @@ def set_config(self, level, entity, key, value, **kw):
+        Returns:
+            exec_cmd response
+        """
+        return self.exec_cmd(cmd=f"ceph config get {entity} {key}", **kw)
```
Same, please move this method out of the RBD class.

sunilangadi2 replied:
Done

manasagowri commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,222 @@
+        int: The return value. 0 for success, 1 for failure.
+    """
+    config = kw.get("config")
+    io = config.get("io_total")
```
Please add default values, if config is not passed by user.

sunilangadi2 replied:
done

manasagowri commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,222 @@
+        if pool["name"] == pool_name:
+            compressed_data = pool["stats"]["compress_bytes_used"]
+            break
+    log.debug(pool["stats"])
```
I think this log.debug should be inside the for loop isn't it?

VasishtaShastry replied:
@sunilangadi2 Also, please add more info to debug message, just printing numbers wouldn't help

sunilangadi2 replied:
done added

VasishtaShastry replied:
okay with log message

manasagowri commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,222 @@
+            break
+    log.debug(pool["stats"])
+    if mode == "compressible":
+        if compressed_data == 0:
```
Are you sure its always 0 and not empty or something?

sunilangadi2 replied:
yes it's always 0 only for uncompressed data refer `'compress_bytes_used': 0`

manasagowri commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,222 @@
+            log.error("Not able to set bluestore_compression_mode to passive")
+            return 1
+
+        log.info("Able to set bluestore_compression_mode to passive")
```
The sentence could be "bluestore_compression_mode set to passive"

sunilangadi2 replied:
modified accordingly

VasishtaShastry commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,222 @@
+    """
+    config = kw.get("config")
+
+    compression_algorithm = config.get("compression_algorithm")
+    compression_mode = config.get("compression_mode")
+    compression_ratio = config.get("compression_ratio")
+
+    pool_name = config[pool_type]["pool"]
+
+    compression_configs = {
+        "compression_algorithm": compression_algorithm,
+        "compression_mode": compression_mode,
+        "compression_required_ratio": compression_ratio,
+    }
```
If we are not using compression_ratio/compression_mode/compression_algorithm after using it to initiate compression_configs, why do we need them ?
In other words, can we merge 
line 97-99 with 104-106

sunilangadi2 replied:
thanks for the code optimization point, Done

sunilangadi2 commented on tests/rbd/rbd_utils.py:
```diff
@@ -773,6 +773,35 @@ def set_config(self, level, entity, key, value, **kw):
         """
         return self.exec_cmd(cmd=f"rbd config {level} set {entity} {key} {value}", **kw)
 
+    def set_ceph_config(self, entity, key, value, **kw):
```
done

sunilangadi2 commented on tests/rbd/test_rbd_compression.py:
```diff
@@ -0,0 +1,133 @@
+
+    Pre-requisites :
+    1. Cluster must be up and running with capacity to create pool
+       (At least with 64 pgs)
```
removed pg info

sunilangadi2 commented:
thanks for the review all addressed all review comments along with merge conflicts,
Test result with latest updated code
[Passed] http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IJ7V4G/
Please review it.

sunilangadi2 commented:
Due to some missing of sequnec of commands during commit 
this PR related commits got messed up 
Raised the new PR with test automation code changes related to it here
https://github.com/red-hat-storage/cephci/pull/2575

Hence closing this PR.
